### ISSUE TITLE
Recompiling at each batch

### ISSUE THREAD

**Issue by ritchieng** [fchollet/keras#4104]
Hi,

I have a Function, call it DetermineLearningRate that takes in the accuracy (and other metrics) at each iteration and gives the learning rate to a simple CNN.

Workflow
1. Simple CNN where DetermineLearningRate() would take in the accuracy and return a learning rate.
2. The TensorFlow Graph of the CNN would take in the learning rate and run 1 iteration and return the accuracy
3. This repeats until all iterations over.

Do I've to recompile 10,000 times if I've 10,000 batches? I tried callbacks and they don't work because the iterations don't wait for my functions to take in the accuracy (and other metrics) and return the learning rate.

**Comment by nixingyang** [comment]
Hi,
Callbacks should work. Do you intend to change the learning rate before each batch. You may inherit the **Callback** class and override the **on_batch_begin** or **on_epoch_begin** functions.
BR.

**Comment by ritchieng** [comment]
@nixingyang Yes I need to change the learning rate before each batch. But this is after the external function gets a metric from the previous batch and returned the learning rate.

The issue I face when I used callbacks (and failed) is that how do I ensure the iteration takes in the new learning rate and not move ahead? I need precise control over the iterations.

An example of what I might do in TF:
1. Build a graph (with the learning rate)
2. Run session with graph for 1 iteration
3. Get the metric from 1 iteration
4. External function would return learning rate
5. Rebuild graph
7. Rerun Session

You get my constraints on why I tried a callback and it doesn't give me this? 

on_batch_end --> get metric
on_batch_begin --> change learning rate
But I can't control precisely as mentioned above because of this external function. Any advice? Might be missing something here.

**Comment by nixingyang** [comment]
Sorry, I don't quite understand your question. You may check the value of the variable **logs** in the **on_batch_begin** and **on_epoch_begin** functions. They should contain the loss value for current batch (at least for the **on_batch_end** function).
